### DISCUSSION TITLE
Homeproxy如何配置可以实现Sing-box的DNS rule的Logical Fields

### DISCUSSION THREAD

**Discussion by laye0619** [immortalwrt/homeproxy#172] [Q&A] [answered]
如题，请教一下如果想实现下面的dns rule，homeproxy应该如何配置？我实在找不到了，多谢！

```
      {
        "type": "logical",
        "mode": "and",
        "rules": [
          {
            "rule_set": "geolocation-!cn",
            "invert": true
          },
          {
            "rule_set": "geoip-cn"
          }
        ],
        "server": "google",
        "client_subnet": "114.114.114.114/24"
      }
```

**Comment by muink** [comment, marked as answer]
目前沒有实现, SB 1.10 有个 inline ruleset 可以直接手写 json
可以用另一种方式实现你说的功能

**Reply by laye0619** [reply]
多谢，我还是期待你们实现吧。

**Reply by 1715173329** [reply]
这种嵌套以目前的架构很难实现，推荐手写 ruleset。

**Reply by laye0619** [reply]
请教一下这种手写ruleset具体怎么操作呢？我需要下载sb 1.10替换openwrt中的sb内核，然后在哪里手写ruleset呢？我知道sing-box配置文件不是都是homeproxy自动生成的吗？那么我应该在homeproxy那个界面手写这种ruleset呢？

十分感谢！！

**Reply by VergilGao** [reply]
> 这种嵌套以目前的架构很难实现，推荐手写 ruleset。

https://github.com/VergilGao/openwrt-sblite  我写了一个 ruleset 的 ui 可以参考一下，就是强行指定一种 ruleset 类型为子规则，因为正经自己写规则也只是为了应付少部分情况，所以强行不让多层嵌套。

**Reply by laye0619** [reply]
> > 这种嵌套以目前的架构很难实现，推荐手写 ruleset。
> 
> https://github.com/VergilGao/openwrt-sblite 我写了一个 ruleset 的 ui 可以参考一下，就是强行指定一种 ruleset 类型为子规则，因为正经自己写规则也只是为了应付少部分情况，所以强行不让多层嵌套。

高级，可惜技术小白不太会用……

**Comment by BedrockDigger** [comment]
> 如题，请教一下如果想实现下面的dns rule，homeproxy应该如何配置？我实在找不到了，多谢！
> 
> ```
>       {
>         "type": "logical",
>         "mode": "and",
>         "rules": [
>           {
>             "rule_set": "geolocation-!cn",
>             "invert": true
>           },
>           {
>             "rule_set": "geoip-cn"
>           }
>         ],
>         "server": "google",
>         "client_subnet": "114.114.114.114/24"
>       }
> ```

我写了一个每日更新的 GitHub Action 来把这个 logical rule 生成为 ruleset，虽然丑陋但能用 XD
https://github.com/BedrockDigger/singbox-enhanced-dns-rule/blob/main/second-last-rule.json
